sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_http_logging_policy.py

```diff
@@ -41,213 +43,303 @@
+                    return
+                log_string = client_settings
+                log_string += db_settings
+                log_string += "\nRequest URL: '{}'".format(redacted_url)
+                log_string += "\nRequest method: '{}'".format(http_request.method)
+                log_string += "\nRequest headers:"
+                for header, value in http_request.headers.items():
+                    value = self._redact_header(header, value)
+                    if value and value != HttpLoggingPolicy.REDACTED_PLACEHOLDER:
+                        log_string += "\n    '{}': '{}'".format(header, value)
+                if isinstance(http_request.body, types.GeneratorType):
+                    log_string += "\nFile upload"
+                    logger.info(log_string, extra=cosmos_logger_attributes)
+                    return
+                try:
+                    if isinstance(http_request.body, types.AsyncGeneratorType):
+                        log_string += "\nFile upload"
+                        logger.info(log_string, extra=cosmos_logger_attributes)
+                        return
+                except AttributeError:
+                    pass
+                if http_request.body:
+                    log_string += "\nA body is sent with the request"
+                    logger.info(log_string, extra=cosmos_logger_attributes)
+                    return
+                log_string += "\nNo body was attached to the request"
+                logger.info(log_string, extra=cosmos_logger_attributes)
+
+            except Exception as err:  # pylint: disable=broad-except
+                logger.warning("Failed to log request: %s", repr(err))
+            return
+        super().on_request(request)
+
+    def on_response(  # pylint: disable=too-many-statements, too-many-branches
+            self,
+            request: PipelineRequest[HTTPRequestType],
+            response: PipelineResponse[HTTPRequestType, HTTPResponseType],
     ) -> None:
-        duration = time.time() - request.context["start_time"] if "start_time" in request.context else None
-        status_code = response.http_response.status_code
-        sub_status_str = response.http_response.headers.get("x-ms-substatus")
-        sub_status_code = int(sub_status_str) if sub_status_str else None
-        verb = request.http_request.method
-        http_version_obj = None
-        url = None
-        if self.diagnostics_handler:
+
+        if self._enable_diagnostics_logging:
+            context = request.context
+            http_response = response.http_response
+            headers = request.http_request.headers
+            sub_status_str = http_response.headers.get("x-ms-substatus")
+            sub_status_code: Optional[int] = int(sub_status_str) if sub_status_str else None
+            url_obj = http_response.internal_response.url  # type: ignore[attr-defined, union-attr]
```

Comment:
I think it might be safer to access the url from the request object. We really should avoid using `internal_response` as this object is unknown. That's not going to cause major problems as this policy already nicely handles all the errors - so worst case scenario is some logs might go missing if you try a different transport - though if there's an easy alternative it might be best.